src/Interpreters/HashJoin.cpp

```diff
@@ -2294,6 +2307,22 @@ void HashJoin::joinBlockImplCross(Block & block, ExtraBlockPtr & not_processed)
                 const IColumn & column_right = *block_right.getByPosition(col_num).column;
                 dst_columns[num_existing_columns + col_num]->insertRangeFrom(column_right, 0, rows_right);
             }
+        };
+
+        if (tmp_stream)
+        {
+            tmp_stream->finishWritingAsyncSafe();
+            auto reader = tmp_stream->getReadStream();
+            while (auto block_right = reader->read())
+            {
+                process_right_block(block_right);
+            }
+        }
+
+        for (const Block & compressed_block_right : data->blocks)
+        {
+            auto block_right = compressed_block_right.decompress();
+            process_right_block(block_right);
```

Comment:
This it not yet finished, right? We should also check a memory limit when we write and here save reader for further calls.